src/Database/LSMTree/Internal/RunReader.hs

```diff
@@ -71,26 +71,41 @@ new ::
 new fs hbio mOffset readerRun = do
     readerKOpsHandle <-
       FS.hOpen fs (runKOpsPath (Run.runRunFsPaths readerRun)) FS.ReadMode
-    -- double the file readahead window (only applies to this file descriptor)
+    -- Double the file readahead window (only applies to this file descriptor)
     FS.hAdviseAll hbio readerKOpsHandle FS.AdviceSequential
 
     (page, entryNo) <- case mOffset of
       Nothing -> do
-        -- load first page from disk, if it exists.
+        -- Load first page from disk, if it exists.
         firstPage <- fromMaybe emptyRawPage <$> readDiskPage fs readerKOpsHandle
         return (firstPage, 0)
 
       Just offset -> do
-        let pageNo = Index.pageSpanStart (Index.search offset index)
-        -- load first page from disk, if it exists.
+        -- Use the index to find the page number for the key (if it exists).
+        let Index.PageSpan pageNo pageEnd = Index.search offset index
+        -- Load first page from disk, if it exists.
         firstPage <- fromMaybe emptyRawPage <$> readDiskPageAt fs pageNo readerKOpsHandle
 
         case rawPageFindKey firstPage offset of
           Just n ->
+            -- Found an appropriate index within the index's page.
             return (firstPage, n)
-          Nothing -> do
-            -- actually, it's the next page
-            nextPage <- fromMaybe emptyRawPage <$> readDiskPageAt fs pageNo readerKOpsHandle
+
+          -- The index said that the key, if it were to exist, would live on pageNo,
+          -- but then rawPageFindKey tells us that in fact
+          -- there is no key greater than or equal to the given offset on this page.
+          -- So what can we conclude from the given information?
+          -- The information tells us that the key does not exist, but that if it were to exist,
+          -- it would be between the last key in this page and the first key in the next page.
+          -- Thus the reader should be initialised to return keys starting from the next page.
+          Nothing | pageNo == pageEnd ->
+            -- If already at the last page, there is no page of content to return.
+            -- Hence, no need to call readDiskPageAt.
+            return (emptyRawPage, 0)
+
+          _ -> do
+            -- Otherwise, we return the beginning of the next page.
+            nextPage <- fromMaybe emptyRawPage <$> readDiskPageAt fs (Index.nextPageNo pageNo) readerKOpsHandle
```

Comment:
So this tries to handle the case where there is no next page, because we were already at the last page (corresponding to asking for a greater than all other keys).

I see that `readDiskPageAt` looks for the `FsReachedEOF` exception. Do we know it definitely works that way when using the "at" functions, like `hGetBufExactlyAt` to request something past the end of the file?

I think it's worth adding a unit test (or a label in a property test) to make sure we cover the case of asking for a key that's past the end of all the keys in the db, so we cover this case here.

Reply:
I would also like to see some tests to be confident about the correctness. It's a bit tricky.

For example, I think `nextPageNo pageNo` here doesn't work if there is a multi-page value. We'd end up loading an overflow page. It should be `nextPageNo pageEnd` instead.

Reply:
One way to get some coverage would be to add `newAtOffset` to the tests in `Test.Database.LSMTree.Internal.RunReaders`.

Reply:
> For example, I think nextPageNo pageNo here doesn't work if there is a multi-page value.

Ah yes, you're right.

Reply:
Ah, I understand better now. I changed the parameter to `Index.nextPageNo pageEnd`.

Just to be certain, it is *unconditionally* `Index.nextPageNo pageEnd`, or do we need to check for a multi-page value and branch between `Index.nextPageNo pageEnd` and `Index.nextPageNo pageNo`?

Reply:
Unconditionally, since for single-page values `pageNo == pageEnd`.